src/tbp/monty/frameworks/models/evidence_matching.py

```diff
@@ -1082,6 +1015,158 @@ def _update_evidence(self, features, displacements, graph_id):
+        num_hypotheses_to_test = hyp_ids_to_test.shape[0]
+        if num_hypotheses_to_test > 0:
+            logging.info(
+                f"Testing {num_hypotheses_to_test} out of "
+                f"{self.evidence[graph_id].shape[0]} hypotheses for "
+                f"{graph_id} "
+                f"(evidence > {evidence_threshold})"
+            )
+
+            # Get evidence update for all hypotheses with evidence > current
+            # _evidence_update_threshold
+            new_evidence = self._calculate_evidence_for_new_locations(
+                graph_id,
+                input_channel,
+                search_locations[hyp_ids_to_test],
+                channel_possible_poses[hyp_ids_to_test],
+                features,
+            )
+            min_update = np.clip(np.min(new_evidence), 0, np.inf)
+
+            # Alternatives (no update to other Hs or adding avg) left in
+            # here in case we want to revert back to those.
+            # avg_update = np.mean(new_evidence)
+            # evidence_to_add = np.zeros_like(channel_hypotheses_evidence)
+            evidence_to_add = np.ones_like(channel_hypotheses_evidence) * min_update
+            evidence_to_add[hyp_ids_to_test] = new_evidence
+
+            # If past and present weight add up to 1, equivalent to
+            # np.average and evidence will be bound to [-1, 2]. Otherwise it
+            # keeps growing.
+            channel_hypotheses_evidence = (
+                channel_hypotheses_evidence * self.past_weight
+                + evidence_to_add * self.present_weight
+            )
+        return search_locations, channel_hypotheses_evidence
+
+    def _replace_hypotheses_in_hpspace(
+        self,
+        graph_id: str,
+        input_channel: str,
+        new_locations_hypotheses: np.ndarray,
+        new_poses_hypotheses: np.ndarray,
+        new_evidence: np.ndarray,
+    ) -> None:
+        """Updates the hypothesis space for a given input channel in a graph.
+
+        This function updates the hypothesis space (for a specific graph and input
+        channel) with a new set of locations, rotations and evidence scores.
+            - If the hypothesis space does not exist for any input channel, a new one
+                is initialized
+            - If the hypothesis space only exists for other channels, a new channel is
+                created with the mean evidence scores of the existing channels
+            - If the hypothesis space exists for the given input channel, the new space
+                replaces the existing hypothesis space
+
+        Args:
+            graph_id (str): The ID of the current graph to update.
+            input_channel (str): Channel's name involved in updating the space
+            new_locations_hypotheses (np.ndarray): New sensor locations hypotheses
+            new_poses_hypotheses (np.ndarray): New object poses hypotheses
```

Comment:
nit: could we call these `new_location_hypotheses` and `new_pose_hypotheses`? (removing the double plural)